apps/vscode/command_client.py

```diff
@@ -1,126 +1,235 @@
-import requests
-import time
+import getpass
+from dataclasses import dataclass
 import json
-from typing import Any
-from talon import Module, actions, app
+import time
 from pathlib import Path
 from tempfile import gettempdir
+from typing import Any, List
+from uuid import uuid4
+
+from talon import Module, actions, Context
 
-is_mac = app.platform == "mac"
+
+STALE_TIMEOUT_MS = 10_000
 
 mod = Module()
 
+ctx = Context()
+mac_ctx = Context()
+
+ctx.matches = r"""
+app: vscode
+"""
+mac_ctx.matches = r"""
+os: mac
+app: vscode
+"""
+
 
 class NotSet:
     def __repr__(self):
         return "<argument not set>"
 
-def run_vscode_command_by_command_palette(command: str):
-    """Execute command via command palette. Preserves the clipboard."""
-    # Clip is noticeably faster than insert
-    if not is_mac:
-        actions.key("ctrl-shift-p")
-    else:
-        actions.key("cmd-shift-p")
 
-    actions.user.paste(f"{command}")
+def run_vscode_command_by_command_palette(command_id: str):
+    """Execute command via command palette. Preserves the clipboard."""
+    actions.user.command_palette()
+    actions.user.paste(command_id)
     actions.key("enter")
 
 
+def write_json_exclusive(path: Path, body: Any):
+    """Writes jsonified object to file, failing if the file already exists
+
+    Args:
+        path (Path): The path of the file to write
+        body (Any): The object to convert to json and write
+    """
+    with path.open("x") as out_file:
+        out_file.write(json.dumps(body))
+
+
+@dataclass
+class Request:
+    command_id: str
+    args: List[Any]
+    wait_for_finish: bool
+    return_command_output: bool
+    uuid: str
+
+    def to_dict(self):
+        return {
+            "commandId": self.command_id,
+            "args": self.args,
+            "waitForFinish": self.wait_for_finish,
+            "returnCommandOutput": self.return_command_output,
+            "uuid": self.uuid,
+        }
+
+
+def write_request(request: Request, path: Path):
+    """Converts the given request to json and writes it to the file, failing if
+    the file already exists unless it is stale in which case it replaces it
+
+    Args:
+        request (Request): The request to serialize
+        path (Path): The path to write to
+
+    Raises:
+        Exception: If another process has an active request file
+    """
+    try:
+        write_json_exclusive(path, request.to_dict())
+    except FileExistsError:
+        stats = path.stat()
```

Comment:
You might want to hoist the rest of this function out of the exception handler, otherwise Python can produce a chained traceback which might not be helpful here. If the write succeeds in the `try` section, just `return`, then the `except FileExistError:` path can just `pass`, and you can put the rest of the code at the outermost indentation level of the function